### PR TITLE
Fixed typos and extended the description of the technical overview

### PR DESCRIPTION
**Pull request by marceloamaral**
What this PR does / why we need it:
The purpose of the PR is to fix some typos and grammar issues, and improve the description of the technical overview in the docs/README.md file.
Which issue(s) this PR fixes:
None
Special notes for your reviewer:
Release note:
```release-note
NONE
```

### CODE REVIEW

**Comment by marceloamaral** [comment]
@dankenigsberg since no one has looked this PR yet, could you please review it?

File: docs/README.md [hunk 1]
```diff
@@ -1,39 +1,34 @@
 ## Technical Overview
 
-Kubernetes allows for extensions to its architecture in the form of custom
-resources: <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/>.
-KubeVirt represents virtual machines as custom resources and manages changes
-to libvirt domains based on the state of those resources.
+Kubernetes allows for extensions to its architecture via *custom resources*, which adds a new endpoint in the Kubernetes API that stores and retrieves a collection API objects of a certain kind. However, the *custom resources* by themselves only enable store and retrieve structured data. To add business logic and specific functionality into Kubernetes, it is necessary to use *custom controllers*, which are clients of the Kubernetes API-Server that typically read an object's `.spec`, possibly do things, and then update the object's `.status`.
```
Comment (dankenigsberg) [inline docs/README.md:3]: Reviewing could be simpler for me if you split this to multiple commits, one for obvious corrections, another for more contested style changes.

For example, I love HREFs, and would like to see them used instead of spelling out the URL

Comment (marceloamaral) [reply docs/README.md]: Thanks! It makes sense, I split to multiple commits and added HREFs

File: docs/README.md [hunk 2]
```diff
@@ -2,27 +2,19 @@
 
 Kubernetes allows for extensions to its architecture in the form of custom
 resources: <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/>.
-KubeVirt represents virtual machines as custom resources and manages changes
-to libvirt domains based on the state of those resources.
+KubeVirt represents virtual machines as custom resources and manages changes to libvirt domains based on the state of those resources.
```
Comment (dankenigsberg) [inline docs/README.md]: I don't understand why you joined these lines. It makes it quite hard to find which typo you have fixed here.

File: docs/README.md [hunk 3]
```diff
@@ -2,27 +2,19 @@
...
- * virt-api: This component provides a HTTP RESTful entrypoint to manage
-   the virtual machines within the cluster.
- * virt-controller: This component manages the state of each VMI within the
-   Kubernetes cluster.
- * virt-handler: This is a daemon that runs on each Kubernetes node. It is
-   responsible for monitoring the state of VMIs according to Kubernetes and
-   ensuring the corresponding libvirt domain is booted or halted accordingly.
- * virt-launcher: This component is a place-holder, one per running VMI. It
-   contains the running VMI and remains running as long as the VMI is defined.
+ * virt-api: This component provides a HTTP RESTful entrypoint to manage the virtual machines within the cluster.
```
Comment (dankenigsberg) [inline docs/README.md]: here, two, I fail to identify why the change is needed.

File: docs/README.md [hunk 4]
```diff
@@ -1,8 +1,12 @@
...
+), which are clients of the Kubernetes API-Server that typically read an object's `.spec`, possibly do things, and then update the object's `.status`.
+
+KubeVirt uses CRDs, *controllers* and other Kubernetes features, to represent and manage traditional virtual machines side by side with containers.
+
+KubeVirt's primary CRD is the VirtualMachine (VM) resource, which contains a collection of VirtualMachineInstance (VMI) objects, which shares similarity with the Pod concept. A VMI represents a single virtualized workload that executes once until completion (i.e., powered off). In addition to the VMI, the key KubeVirt components are the virt-api, the virt-controller, the virt-handler, and the virt-launcher.
```
Comment (dankenigsberg) [inline docs/README.md:8]: I'm not sure at all that this is factually correct. One VM is related to a single VMI at most.

File: docs/README.md [hunk 5]
```diff
@@ -10,10 +10,10 @@ KubeVirt's primary CRD is the VirtualMachine (VM) resource, which contains a col
 
 ### Project Components
 
- * virt-api: This component provides a HTTP RESTful entrypoint to manage the virtual machines within the cluster.
- * virt-controller: This component manages the state of each VMI within the Kubernetes cluster.
- * virt-handler: This is a daemon that runs on each Kubernetes node. It is responsible for monitoring the state of VMIs according to Kubernetes and ensuring the corresponding libvirt domain is booted or halted accordingly.
- * virt-launcher: This component is a place-holder, one per running VMI. It contains the running VMI and remains running as long as the VMI is defined.
+* **virt-api**: This component provides an HTTP RESTful entry-point to manage the virtual machines within the cluster. It is created by adding the KubeVirt's CRDs.
```
Comment (dankenigsberg) [inline docs/README.md:15]: I'm not sure if the added sentence is useful. It may confuse the reader - what is created by CRDs, the entry point? or virt-api?

File: docs/README.md (hunk 2 above)
Comment (marceloamaral) [reply docs/README.md]: I see, let me revert that

File: docs/README.md (hunk 4 above)
Comment (marceloamaral) [reply docs/README.md:8]: I see, since [CRD documentation](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) says that it is a resource that contains a collection of objects, I added that here
But, if one VM resource is related to a single VMI at most, IMHO i think is worth mentioning that here!

File: docs/README.md (hunk 5 above)
Comment (marceloamaral) [reply docs/README.md:15]: Right, I will omit that

File: docs/README.md (hunk 2 above)
Comment (marceloamaral) [reply docs/README.md]: The typo was actually regarding the word  `enviornment --> environment`

File: docs/README.md (hunk 3 above)
Comment (marceloamaral) [reply docs/README.md]: ok, i did one commit for the changes of each component description....

File: docs/README.md [hunk 6]
```diff
@@ -1,9 +1,27 @@
 ## Technical Overview
 
-Kubernetes allows for extensions to its architecture in the form of custom
-resources: <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/>.
-KubeVirt represents virtual machines as custom resources and manages changes
-to libvirt domains based on the state of those resources.
+Kubernetes allows for extensions to its architecture via
+[*custom resources*]( https://Kubernetes.io/docs/concepts/extend-Kubernetes/api-extension/custom-resources/),
+which add a new endpoint in the Kubernetes API that stores and retrieves a
+collection API objects of a certain kind. However, the *custom resources*
```
Comment (dankenigsberg) [inline docs/README.md:6]: I find the "However" sentence mostly repetitive. I think you should just explain why controllers are needed, not why CRDs are not enough.

File: docs/README.md [hunk 7]
```diff
@@ -1,9 +1,27 @@
...
+which are clients of the Kubernetes API-Server that typically read an
+object's `.spec`, possibly do things, and then update the object's
+`.status`.
+
+KubeVirt uses CRDs, *controllers* and other Kubernetes features, to
+represent and manage traditional virtual machines side by side with
+containers.
+
+KubeVirt's primary CRD is the VirtualMachine (VM) resource, which contains
+a collection of VirtualMachineInstance (VMI) objects, which shares
```
Comment (dankenigsberg) [inline docs/README.md]: Again, I don't think it is a collection.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 7 above)
Comment (marceloamaral) [reply docs/README.md]: I fixed that in the last commit.. Well, I think now I have too many commits, sorry for that!

File: docs/README.md (hunk 6 above)
Comment (marceloamaral) [reply docs/README.md:6]: Makes sense, let me fix that!

File: docs/README.md [hunk 8]
```diff
@@ -1,21 +1,45 @@
...
+ objects are submitted to the Kubernetes API-Server, this controller takes
+ notice and creates the pod in which the VM will run and delegates the
+ other management operations to the *virt-handler* component.
+ * **virt-handler**: This is a daemon that runs on each Kubernetes node. It is
    responsible for monitoring the state of VMIs according to Kubernetes and
-   ensuring the corresponding libvirt domain is booted or halted accordingly.
- * virt-launcher: This component is a place-holder, one per running VMI. It
-   contains the running VMI and remains running as long as the VMI is defined.
+   ensuring the corresponding libvirt domain is booted or halted accordingly. To perform these operations, the virt-handler signals and send the VM's CRD object to the *virt-launcher* component.
+ * **virt-launcher**: This component is a place-holder, one per running VMI. It
```
Comment (jean-edouard) [inline docs/README.md:40]: What do you mean by place-holder? To me that sentence sounds like virt-launcher sits there and does nothing, when it's where all the virtualization actually happens.

**Review by jean-edouard — COMMENTED** [review]
/approve
Just one question in-line.
Also, thanks a lot for improving this doc, very useful info!

File: docs/README.md (hunk 8 above)
Comment (marceloamaral) [reply docs/README.md]: Thanks @jean-edouard! I am sorry for the late reply, I misunderstood the messages and I thought this PR was merged....
Anyway, I did an update accordingly to your comment!

File: docs/README.md [hunk 9]
```diff
@@ -1,21 +1,45 @@
...
+KubeVirt uses CRDs, *controllers* and other Kubernetes features, to
+represent and manage traditional virtual machines side by side with
+containers.
+
+KubeVirt's primary CRD is the VirtualMachine (VM) resource, which contains
+a VirtualMachineInstance (VMI) object. The VMI
+shares similarity with the Pod concept and represents a single virtualized
+workload that executes once until completion (i.e., powered off). In
+addition to the VMI, the key KubeVirt components are the virt-api, the
+virt-controller, the virt-handler, and the virt-launcher.
```
Comment (davidvossel) [inline docs/README.md:24]: The VMI isn't exactly a component like virt-api, virt-controller, virt-handler, virt-launcher.  The VMI is an API that components interact with.

File: docs/README.md [hunk 10]
```diff
@@ -1,21 +1,45 @@
...
+object's `.spec`, possibly do things, and then update the object's
+`.status`.
+
+KubeVirt uses CRDs, *controllers* and other Kubernetes features, to
+represent and manage traditional virtual machines side by side with
+containers.
+
+KubeVirt's primary CRD is the VirtualMachine (VM) resource, which contains
+a VirtualMachineInstance (VMI) object. The VMI
+shares similarity with the Pod concept and represents a single virtualized
```
Comment (davidvossel) [inline docs/README.md:21]: The closest thing we have to a k8s analogy for the VM->VMI relationship is a k8s StatefulSet of replica 1.  The StatefulSet contains a pod template and manages the lifecycle of the corresponding pod. The VM contains a VMI template and manages the corresponding lifecycle of the VMI.

File: docs/README.md [hunk 11]
```diff
@@ -1,21 +1,45 @@
...
+ * **virt-api**: This component provides a HTTP RESTful entrypoint to manage
    the virtual machines within the cluster.
- * virt-controller: This component manages the state of each VMI within the
-   Kubernetes cluster.
- * virt-handler: This is a daemon that runs on each Kubernetes node. It is
+ * **virt-controller**: This component is a Kubernetes Operator that
+ manages the state of each VMI within the Kubernetes cluster. When new VM
+ objects are submitted to the Kubernetes API-Server, this controller takes
+ notice and creates the pod in which the VM will run and delegates the
+ other management operations to the *virt-handler* component.
```
Comment (davidvossel) [inline docs/README.md:36]: it seems like there's some mixed usage of VMI and VM here. Those terms are very specific.  When a VM object is submitted to the cluster, it results in a VMI object being created. That VMI object results in a corresponding pod.  The virt-controller component itself is made of multiple microservices (controller loops) internally that have their own separation of concerns. 

also, the term "operator" is thrown in here when in the technical overview the term "controller" is used instead. 

As an overview of virt-controller, i wouldn't necessarily get into all these details.  Simply saying `This component is a Kubernetes controller that manages the lifecycle of virtual machines within the Kubernetes cluster` might be enough.

File: docs/README.md [hunk 12]
```diff
@@ -1,21 +1,45 @@
...
+ * **virt-handler**: This is a daemon that runs on each Kubernetes node. It is
    responsible for monitoring the state of VMIs according to Kubernetes and
-   ensuring the corresponding libvirt domain is booted or halted accordingly.
- * virt-launcher: This component is a place-holder, one per running VMI. It
-   contains the running VMI and remains running as long as the VMI is defined.
+   ensuring the corresponding libvirt domain is booted or halted accordingly. To perform these operations, the virt-handler signals and send the VM's CRD object to the *virt-launcher* component.
```
Comment (davidvossel) [inline docs/README.md:39]: technically we're sending the VMI spec to virt-launcher, but it's a little more complex than that. Theres a grpc channel established between virt-handler and each virt-launcher that allows commands to be sent from virt-handler to virt-launcher. Within those commands, the VMI spec is sometimes passed. 

all that said. Simply saying something like "virt-handler has a communication channel with each virt-launcher that is used to manage the lifecycle of the qemu process within the virt-launcher pod" might be enough

File: docs/README.md [hunk 13]
```diff
@@ -1,21 +1,45 @@
...
+ * **virt-launcher**: There is one per running VMI. It
+   contains the running VMI and remains running as long as the VMI is defined. It uses a local libvirtd instance within its container to
+   start or terminate the VM.
```
Comment (davidvossel) [inline docs/README.md:42]: virt-launcher doesn't remain running as long as the VMI is defined, it remains running as long as we qemu process is running. The VMI object might still exist in the cluster in a terminated state indefinitely after that point which doesn't indicate a corresponding pod is present.

Perhaps saying "This component directly manages the lifecycle of the qemu process within the VMI's pod and receives lifecycle commands from virt-handler" would be more accurate.

**Review by davidvossel — COMMENTED** [review]
thanks for taking the time to update this. I found a few inconsistencies from a technical perspective and made a few comments on how we might clear it up.

File: docs/README.md (hunk 9 above)
Comment (marceloamaral) [reply docs/README.md]: Fixed that!

File: docs/README.md (hunk 10 above)
Comment (marceloamaral) [reply docs/README.md]: Makes sense, I simplified that.

File: docs/README.md (hunk 11 above)
Comment (marceloamaral) [reply docs/README.md]: Fixed!

File: docs/README.md (hunk 12 above)
Comment (marceloamaral) [reply docs/README.md]: fixed

File: docs/README.md (hunk 13 above)
Comment (marceloamaral) [reply docs/README.md]: fixed :)

Comment (marceloamaral) [reply docs/README.md]: @davidvossel could you please review the updates?

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

**Comment by marceloamaral** [comment]
\retest
